rfcs/ALL/Access Control Rules.md

```diff
@@ -0,0 +1,133 @@
+
+Internally, for each combination of *interface* + *flow* + *action*  (example:`l0`+`egress`+`put`) in the rules vector, we construct *allow* and *deny* KeTrees (key-expression tries). The *allow* KeTree is built using all the key-expressions provided in the config on which that *interface* is allowed to do that *action* on a particular *flow*. On receiving an authorization request, the key-expression in the request is matched against the appropriate KeTrees to confirm authorization.
+
+## Priority 
+
+For each message, the access control checks are made in the following order of priority:
+*explicit deny* rule > *explicit allow* rule > *default_permission* rule
+
+1. Explicit rules with deny permissions are given topmost priority. This means that if the key-expression in the  authorization request matches against anything in the *deny* KeTree, it will be denied immediately. If there is no match, then the *explicit allow* rules are checked.
+2. Explicit rules with allow permissions come next, i.e., if request matches anything in the *allow* tree then it will be allowed.
+3. The *default_permission* value has least priority. It is applied only if there are no matches in the previous two steps.
+    
+    Note: if *default_permission* is set to `allow`, then there is no need for setting explicit allow rules, since they will be overlooked by the matcher anyways. The same logic does not apply if *default_permission* is set to `deny` and explicit deny, due to permission priorities (see pseudo-code below for more detail)"
+
+The decision logic is explained by the following pseudocode:
+
+```rust
+fn is_allowed(key_expr) -> decision {
+   if default_permission == deny {
+      if !deny_ketree.match(key_expr) && allow_ketree.match(key_expr) {
+            return true;
+      } else {
+          return false;
+      }
+   } else { 
+      // default_permission = allow
+      if deny_ketree.match(key_expr) {
+         return false;
+      } 
+      // no need to check for allow_ketree here
+      return true; 
+   }
+}
+```
+
+## Key-Expression Matching
+
+All requets are macthed on keys and key expressions (for more information on key expressions, check out [Key_Expressions](https://github.com/eclipse-zenoh/roadmap/blob/main/rfcs/ALL/Key%20Expressions.md)) . Therefore, it is important to understand how the key-expression matching works, since it ultimately decides the behavior of the access control logic. In matching a key-expression against a KeTree, it will match as a positive only if it is *included in* (*equal to* or *subset of*) the key-expressions specified in the KeTree. A partial match or being a superset will not result in a match.
```

Comment:
```suggestion
All requests are macthed on keys and key expressions (for more information on key expressions, check out [Key_Expressions](https://github.com/eclipse-zenoh/roadmap/blob/main/rfcs/ALL/Key%20Expressions.md)) . Therefore, it is important to understand how the key-expression matching works, since it ultimately decides the behavior of the access control logic. In matching a key-expression against a KeTree, it will match as a positive only if it is *included in* (*equal to* or *subset of*) the key-expressions specified in the KeTree. A partial match or being a superset will not result in a match.
```